D3D11 video driver

Milek7 opened the pull request:
## Motivation / Problem

OpenGL experience on Windows might be subpar, fresh Windows installations might not have OpenGL drivers installed, etc.

## Description

Mostly port of the OpenGL driver.

## Limitations

- Feature level 10_0 required. Might be possible to get lower with slightly more code.

- ~Fullscreen not yet implemented.~
Generally DXGI model doesn't play too well with existing code. ~What win32 does currently, recreating the window for each mode switch is completely unnecessary, in DXGI we can resize/switch modes without recreating window or even swapchain. However to do that requires rework of win32 base driver window creation code.~ Moreover, we might want to ditch exclusive fullscreen mode entirely and switch to borderless fullscreen windows. On modern compositors exclusive fullscreen have no benefits

- Exclusive fullscreen is implemented, but on purpose it is not allowed to change system resolution. It can be changed by adding `DXGI_SWAP_CHAIN_FLAG_ALLOW_MODE_SWITCH` to swapchain creation flags, but I'm not sure if this is desired.

- ~I had to align staging texture width to 128 pixels, otherwise pitch gets screwed, I have no idea why.~ Fixed.

- I didn't implement sprite cursor drawing.

- Might not be that necessary after all, as during development of this I found bug that likely caused most of the OpenGL crashes (fixed in #9100)

- ~Codestyle might be off.~

orudge commented on src/os/windows/comptr.h:
```diff
@@ -0,0 +1,91 @@
+
+ /** @file comptr.h COM object wrapper. */
+
+#ifndef COMPTR_H
```
FWIW, src/sound/xaudio2_s.cpp makes use of #include <wrl/client.h> and the Microsoft::WRL::ComPtr class. This may have a minimum Windows SDK version, but given we use it elsewhere in OpenTTD, it doesn't seem unreasonable to me to use it here too.

Milek7 replied:
Oh, from all MS variants of COM smart pointer I didn't see that one. I think mine was a bit nicer to use, but I replaced it now with WRL one.

michicc commented:
I haven't tried it, but two general comments after a short look through the diff:

A driver error on init should gracefully fall back to another driver, which means all those `throw`s are not good.

Also, the D3DCompiler DLL exists in various versions among OS versions and stuff. Official MS recommendations is the pre-compile shaders with `fxc` into byte code instead of using the compiler, otherwise you might have to make sure the proper DLL is packaged together with the exe.

Milek7 committed:
Codechange: Allow different pitch for animation buffer

Milek7 commented:
> A driver error on init should gracefully fall back to another driver, which means all those `throw`s are not good.

I removed all throws, though I don't think they mattered much as all were after context creation (with proper feature level) so failures weren't expected.


> Also, the D3DCompiler DLL exists in various versions among OS versions and stuff. Official MS recommendations is the pre-compile shaders with `fxc` into byte code instead of using the compiler, otherwise you might have to make sure the proper DLL is packaged together with the exe.

I'm not sure about adding build-time dependency on fxc, is it available eg. on MinGW builds?
d3dcompiler 47 is included since win8.1 as a part of system. For now I changed d3dcompiler loading function to search for all the versions down to 40, which should cover any installed directx runtimes since 2008.

I implemented exclusive fullscreen, but on purpose it is not allowed to change system resolution. It can be changed by adding DXGI_SWAP_CHAIN_FLAG_ALLOW_MODE_SWITCH to swapchain creation flags, but I'm not sure if this is desired. (note that recent Windows versions might transform it into fullwindow, so it won't be true exclusive fullscreen anyway.)

Milek7 committed:
Feature: D3D11 video driver
